src_lockfree/mpsc_queue.ml

```diff
@@ -1,152 +1,145 @@
-    (* CAS failed because [p] is no longer the tail (or is already closed). *)
-    Node.fold (Atomic.get p.next)
-      ~none:(fun () -> assert false)
-        (* Can't switch from another state to [none] *)
-      ~some:(fun p_next ->
-        (* Make [tail] more up-to-date if it hasn't changed already *)
-        ignore (Atomic.compare_and_set t.tail p p_next : bool);
-        (* Retry *)
-        close t)
+and ('a, _) tail =
+  | Snoc : 'a tail_pack * 'a -> ('a, [> `Snoc ]) tail
+  | Topen : ('a, [> `Topen ]) tail
+  | Tclosed : ('a, [> `Tclosed ]) tail
 
-let pop_opt t =
-  let p = t.head in
-  (* [p] is the previously-popped item. *)
-  let node = Atomic.get p.next in
-  Node.fold node
-    ~none:(fun () -> None)
-    ~some:(fun node ->
-      t.head <- node;
-      let v = node.value in
-      node.value <- Obj.magic ();
-      (* So it can be GC'd *)
-      Some v)
+and 'a tail_pack = T : ('a, [< `Snoc | `Topen | `Tclosed ]) tail -> 'a tail_pack
+[@@unboxed]
 
-exception Empty
+let create () =
+  let tail = Multicore_magic.copy_as_padded @@ Atomic.make (T Topen) in
+  let head = Multicore_magic.copy_as_padded @@ H Hopen in
+  Multicore_magic.copy_as_padded { tail; head }
+
+(* let push t x = t.tail <- T (Snoc (t.tail, x)) *)
+
+let rec push t x =
+  match Atomic.get t.tail with
+  | T Tclosed -> raise Closed
+  | before ->
+      let after = T (Snoc (before, x)) in
+      if not (Atomic.compare_and_set t.tail before after) then push t x
+
+(*  *)
+let rec rev_to (head : (_, [< `Cons ]) head) = function
+  | T Topen -> head
+  | T Tclosed -> assert false
+  | T (Snoc (xs, x)) -> rev_to (Cons (x, H head)) xs
+
+let rec rev (head) = function
+  | T Topen -> head
+  | T Tclosed -> assert false
+  | T (Snoc (xs, x)) -> rev (H (Cons (x, head))) xs
+
+let rev_pop = function
+  | (Snoc (tail, x) : (_, [< `Snoc ]) tail) -> rev_to (Cons (x, H Hopen)) tail
+
+let rec append a b =
+  match b with
+  | H Hclosed ->
+      failwith
+        "This cannot happen, maybe you are running close concurrently with \
+         another [close]"
+  | H Hopen -> a
+  | H (Cons (x, xs)) -> H (Cons (x, append a xs))
 
-let pop t =
-  let p = t.head in
-  (* [p] is the previously-popped item. *)
-  let node = Atomic.get p.next in
-  Node.fold node
-    ~none:(fun () -> raise Empty)
-    ~some:(fun node ->
-      t.head <- node;
-      let v = node.value in
-      node.value <- Obj.magic ();
-      (* So it can be GC'd *)
-      v)
+let pop_opt t =
+  match t.head with
+  | H Hclosed -> raise Closed (*Closed and empty*)
+  | H (Cons (x, xs)) ->
+      t.head <- xs;
+      Some x
+  | H Hopen -> (
+      match Atomic.exchange t.tail (T Topen) with
+      | T Topen -> None
+      | T Tclosed ->
+          failwith
+            "This cannot happen, maybe you are running close concurrently with \
+             [pop]"
+      | T (Snoc _ as tail) -> (
+          match rev_pop tail with
+          | Cons (x, xs) ->
+              t.head <- xs;
```

Comment:
Here (and in `pop`) performance could be improved slightly by skipping the write `t.head <- xs` in case `xs == H Hopen`, e.g.

```diff
          match rev_pop tail with
+         | Cons (x, H Hopen) -> Some x
          | Cons (x, xs) ->
              t.head <- xs;
```

Writing to a potentially pointer containing field in OCaml is relatively expensive as it requires a write barrier.